horovod/common/response_cache.cc

```diff
@@ -0,0 +1,223 @@
+    auto& entry = cache_.back().first;
+    cache_bit = tensor_name_to_bit_[entry.tensor_names()[0]];
+    tensor_name_to_bit_.erase(entry.tensor_names()[0]);
+    cache_.pop_back();
+    cache_.push_front(std::make_pair(response, std::move(params)));
+  } else {
+    cache_bit = cache_iters_.size();
+    cache_iters_.resize(cache_bit + 1);
+    cache_.push_front(std::make_pair(response, std::move(params)));
+  }
+
+  cache_iters_[cache_bit] = cache_.begin();
+  tensor_name_to_bit_[response.tensor_names()[0]] = cache_bit;
+
+  bits_outdated_ = true;
+}
+
+void ResponseCache::put(const Response& response, TensorTable& tensor_table) {
+  if (capacity_ == 0) {
+    return;
+  }
+
+  // If response is fused, split back into individual responses
+  if (response.tensor_names().size() > 0) {
+    for (auto& name : response.tensor_names()) {
+      Response new_response;
+      new_response.add_tensor_name(name);
+      new_response.set_response_type(response.response_type());
+      new_response.set_devices(response.devices());
+      new_response.set_tensor_sizes(response.tensor_sizes());
+
+      // Populate tensor parameters from tensor_table entry
+      auto& tensor_entry = tensor_table[name];
+      TensorParams params;
+      params.device = tensor_entry.device;
+      params.dtype = tensor_entry.tensor->dtype();
+      params.shape = tensor_entry.tensor->shape().to_vector();
+
+      this->put_(new_response, params);
+    }
+  } else {
+    auto& tensor_entry = tensor_table[response.tensor_names()[0]];
+    TensorParams params;
+    params.device = tensor_entry.device;
+    params.dtype = tensor_entry.tensor->dtype();
+    params.shape = tensor_entry.tensor->shape().to_vector();
+
+    this->put_(response, params);
+  }
+}
+
+const Response& ResponseCache::get_response(const Request& message) {
+  assert(this->cached(message));
+  uint32_t cache_bit = tensor_name_to_bit_[message.tensor_name()];
+  auto it = cache_iters_[cache_bit];
+  cache_.push_front(std::move(*it));
+  cache_.erase(it);
+  cache_iters_[cache_bit] = cache_.begin();
+  bits_outdated_ = true;
+  return cache_.front().first;
+}
+
+const Response& ResponseCache::get_response(uint32_t cache_bit) {
+  assert(cache_bit < cache_iters_.size());
+  auto it = cache_iters_[cache_bit];
+  cache_.push_front(std::move(*it));
+  cache_.erase(it);
+  cache_iters_[cache_bit] = cache_.begin();
+  bits_outdated_ = true;
+  return cache_.front().first;
+}
+
+const Response& ResponseCache::peek_response(const Request& message) const {
+  assert(this->cached(message));
+  uint32_t cache_bit = tensor_name_to_bit_.at(message.tensor_name());
+  return std::get<0>(*cache_iters_[cache_bit]);
+}
+
+const Response& ResponseCache::peek_response(uint32_t cache_bit) const {
+  assert(cache_bit < cache_iters_.size());
+  return std::get<0>(*cache_iters_[cache_bit]);
+}
+
+uint32_t ResponseCache::peek_cache_bit(const Request& message) const {
+  assert(this->cached(message));
```

Comment:
Add assert that bits are not outdated?

Reply:
An assert on `bits_outdated_` would not be correct here. This is because in the case that invalid elements are erased and `erase_response` is called, this flag is set to be true. However, we still require extracting response information on the bits we collected before erasing. See https://github.com/horovod/horovod/blob/11a5d6d365960efd90374144cffb83d558e259e1/horovod/common/operations.cc#L1411.